crates/pop-parachains/src/call/metadata/params.rs

```diff
@@ -0,0 +1,256 @@
+		if let Some(sub_type_id) = type_info.type_params.first().and_then(|param| param.ty) {
+			// Recursive for the sub parameters
+			let sub_param = type_to_param(name.clone(), registry, sub_type_id.id)?;
+			Ok(Param {
+				name,
+				type_name: sub_param.type_name,
+				sub_params: sub_param.sub_params,
+				is_optional: true,
+				is_tuple: false,
+				is_variant: false,
+			})
+		} else {
+			Err(Error::MetadataParsingError(name))
+		}
+	} else {
+		// Determine the formatted type name.
+		let type_name = format_type(type_info, registry);
+		match &type_info.type_def {
+			TypeDef::Primitive(_) |
+			TypeDef::Array(_) |
+			TypeDef::Sequence(_) |
+			TypeDef::Compact(_) => Ok(Param {
+				name,
+				type_name,
+				sub_params: Vec::new(),
+				is_optional: false,
+				is_tuple: false,
+				is_variant: false,
+			}),
+			TypeDef::Composite(composite) => {
+				let sub_params = composite
+					.fields
+					.iter()
+					.map(|field| {
+						// Recursive for the sub parameters of composite type.
+						type_to_param(
+							field.name.clone().unwrap_or(name.clone()),
+							registry,
+							field.ty.id,
+						)
+					})
+					.collect::<Result<Vec<Param>, Error>>()?;
+
+				Ok(Param {
+					name,
+					type_name,
+					sub_params,
+					is_optional: false,
+					is_tuple: false,
+					is_variant: false,
+				})
+			},
+			TypeDef::Variant(variant) => {
+				let variant_params = variant
+					.variants
+					.iter()
+					.map(|variant_param| {
+						let variant_sub_params = variant_param
+							.fields
+							.iter()
+							.map(|field| {
+								// Recursive for the sub parameters of variant type.
+								type_to_param(
+									field.name.clone().unwrap_or(variant_param.name.clone()),
+									registry,
+									field.ty.id,
+								)
+							})
+							.collect::<Result<Vec<Param>, Error>>()?;
+						Ok(Param {
+							name: variant_param.name.clone(),
+							type_name: "".to_string(),
+							sub_params: variant_sub_params,
+							is_optional: false,
+							is_tuple: false,
+							is_variant: true,
+						})
+					})
+					.collect::<Result<Vec<Param>, Error>>()?;
+
+				Ok(Param {
+					name,
+					type_name,
+					sub_params: variant_params,
+					is_optional: false,
+					is_tuple: false,
+					is_variant: true,
+				})
+			},
+			TypeDef::Tuple(tuple) => {
+				let sub_params = tuple
+					.fields
+					.iter()
+					.enumerate()
+					.map(|(index, field_id)| {
+						type_to_param(
+							format!("Index {index} of the tuple {name}"),
+							registry,
+							field_id.id,
+						)
+					})
+					.collect::<Result<Vec<Param>, Error>>()?;
+
+				Ok(Param {
+					name,
+					type_name,
+					sub_params,
+					is_optional: false,
+					is_tuple: true,
+					is_variant: false,
+				})
+			},
+			_ => Err(Error::MetadataParsingError(name)),
+		}
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use crate::set_up_client;
+	use anyhow::Result;
+
+	#[tokio::test]
+	async fn field_to_param_works() -> Result<()> {
```

Comment:
I agree with you. In this specific function, the two are closely related since the second was created to be used recursively. It seems more logic to combine them into a single test.